### ISSUE TITLE
[enhancement, question, issue] Key and value not the same size results in RuntimeError

### ISSUE THREAD

**Issue by strohhut22** [atom-moyer/getpy#7]
Hello, thank you a lot for you work on this implementation.  The hashmap is excellent, simple, and fast! I am trying to use it right now to speed up some computation where I need fast access to values of many keys in a large dictionary.

However, when using keys and values with different shapes, e.g., shape (85770, 3) for the keys and values of any shape (85770, X) with X > 3 or X < 3, the result is a RuntimeError. I kindly want to ask why it is the case that keys and values MUST be of the same shape/size.

Example to reproduce:
```
  ### keys and vals
  keys = np.ones( (85770, 3)).astype(np.dtype('u8'))
  vals =  np.ones( (85770, 10)).astype(np.dtype('u8'))
  print(keys.shape, vals.shape)
  >> ((85770, 3), (85770, 10))

  gp_dict = gp.Dict(np.dtype('u8'), np.dtype('u8'), default_value=np.zeros(10))
  gp_dict[keys] = vals

  >> RuntimeError: The size of the key and value must match. 
  ---------------------------------------------------------------------------
  RuntimeError                              Traceback (most recent call last)
  <ipython-input-229-160a4522ca76> in <module>
        3 
        4 gp_dict = gp.Dict(np.dtype('u8'), np.dtype('u8'), default_value=np.zeros(10)))
  ----> 5 gp_dict[keys] = vals
  
  /opt/conda/lib/python3.8/site-packages/getpy/getpy.py in __setitem__(self, key, value)
       34     def __setitem__(self, key, value):
       35         if self.__safe_mode: assert key.dtype == self.__key_type and value.dtype == self.__value_type, (key.dtype, value.dtype)
  ---> 36         self.__dict.__setitem__(key, value)
       37 
       38 
  
  RuntimeError: The size of the key and value must match.
 ```

I see that both to have the same shape is enforced in src/getpy/getpy.cpp in __setitem__ (line 62).
```
( key_array.size() != value_array.size() )
            throw std::runtime_error("The size of the key and value must match.");:
```
When querying the values of keys, the shape for the output array's shape is copied from the query's shape (key shape) (line 44/45). I see this as the reason why it is enforced. But I do not understand why it is denied in the first place to have different shapes. I assume the shape of the values could be stored in some different way. 
```
auto result_array = py::array_t<Value> ( key_array.request().shape );
        auto * result_array_ptr = (Value *) result_array.request().ptr;
```

I am not an expert in this, so of course I do not fully understand how the c++ implementation of the parallel-hashmap works. Maybe I am missing something really obvious, but I do not see why it would not technically be possible to store different shapes for the values, and access them in a parallel way. They are just stored at some address in memory and the hashmap should just query by the respective hashed index?

**Comment by atom-moyer** [comment]
Ah. Ok there is a misunderstanding. 

`gp_dict = gp.Dict(np.dtype('u8'), np.dtype('u8'), default_value=np.zeros(10))`

This line does not work how you are expecting. Specifically, you cannot store 10 u8's per key. When you make that init call, you are instantiating the pre-compiled hash table with 64 bit keys and 64 bit values, so whatever you can fit into 64 bits is fair game.

The vectorized call that you are doing when X=3 is actually assigning 85770 * 3 keys to 85770 * 3 values. You can think that it is just flattening the arrays, but it is more convenient. 

Does that make sense?

What you are looking for is a MultiDict, which assigns arrays/sets of the value type. I have implemented that, with less documentation. Specifically, the MultiDict that I have written is a Dictionary(key, Set(value)), so you cannot store duplicate values. Adding a MultiDict_not_unique(key, Array(value, N)) could be possible, but usually I just use bit packing and S16 dtype to get that behavior. There are trade offs between the various implementations. Having all of the would be best, but in my work flows, this works best for me.

Note, instead of returning a numpy matrix that you might expect, the MultiDict will return a python list of numpy arrays because the value vectors are not guaranteed to be the same size.

**Comment by atom-moyer** [comment]
With that said, I totally understand why you thought it should work that way. It would be very cool if it did. 

If it worked that way, it would require me to compile A LOT (1000x) of different dictionaries due to the combinatorics of the key/value pairs (compilation already requires 16gb and 1 hr if I remember correctly). Also, I would have to add some magic under the hood to auto-detect your array sizes, which I think is solvable. I have tried to avoid doing stuff like that though.

If you have a specific pair of data types that you really need and you are not able to fit it into an S16 (128 bits), then I could help you compile with the additional hash table.

Basically, you need to add another template to the file: `src/getpy_types.cpp`
Then, tie that into the python in the file: `getpy/getpy_types.py`

**Comment by strohhut22** [comment]
Thank you a lot for the fast and detailed reply!

> With that said, I totally understand why you thought it should work that way. It would be very cool if it did. If it worked that way, it would require me to compile A LOT (1000x) of different dictionaries due to the combinatorics of the key/value pairs (compilation already requires 16gb and 1 hr if I remember correctly).

I see, thank you for the detailed explanation! 

But it can still be done with the cp.Dict what I need: I actually found a quick and dirty workaround that at least for me is suitable. Padding zeros to the end of my key vector makes both key and value the same size... And it queries lightspeed faster than looping through everything with a standard dict... and I am happy. Fits in 64 bit what I store in both sides.


> If you have a specific pair of data types that you really need and you are not able to fit it into an S16 (128 bits), then I could help you compile with the additional hash table.

Thanks, but I am not in need for that right now. What I need fits perfectly fine with your cp.Dict, I just really want to query 3d coordinates and have a bunch of 0s and 1s as values that can be duplicates + they need to be in the same order, that is why the MultiDict with Set as value does not work for me.

**Comment by atom-moyer** [comment]
From my understanding of your intentions, I do not think that will work the
way you are intending.

You cannot use a vector as a key or value. You can only store the exact
data type that you assign at init, NOT a vector of the dtype. The
vectorized interface that you are using is streamlining inserting into the
dictionary.

If you pad with zeros, you are inserting 85770 * 10 individual key/value
pairs, NOT 85770 individual key/value pairs. 7/10ths of your inserts are
going to overwrite the 0 key value.

To do what you want to do, you need to bitpack the XYZ coordinates to an
S16 using this datatype (0[int32], x[int32], y[int32], z[int32]) using the
nparray.view() function. Then you will need to store all of your values as
bitpacked bool arrays.

**Comment by strohhut22** [comment]
Thanks for you fast and detailed reply!
I just thought I had a small workaround by just padding zeros to the end of my key. But it does not seem to work, and you explained why. 
> What you are looking for is a MultiDict, which assigns arrays/sets of the value type. I have implemented that, with less documentation. Specifically, the MultiDict that I have written is a Dictionary(key, Set(value)), so you cannot store duplicate values. Adding a MultiDict_not_unique(key, Array(value, N)) could be possible, but usually I just use bit packing and S16 dtype to get that behavior.  

This is a really good tip, thanks! I think I will try bit packing, I have a fixed value of length N which is just a bunch of 0s and 1s. So would be just a bit string of length N. At most probably 20 bit for me. I will try this out, thanks a lot. :)


> With that said, I totally understand why you thought it should work that way. It would be very cool if it did.

> If it worked that way, it would require me to compile A LOT (1000x) of different dictionaries due to the combinatorics of the key/value pairs (compilation already requires 16gb and 1 hr if I remember correctly). Also, I would have to add some magic under the hood to auto-detect your array sizes, which I think is solvable. I have tried to avoid doing stuff like that though.

Oof, this is crazy. I think the easiest way to detect the array size of the values would be to just use it as parameter at dictionary creation. Would limit to same size for each value.

**Comment by strohhut22** [comment]
> You cannot use a vector as a key or value. You can only store the exact
data type that you assign at init, NOT a vector of the dtype. The
vectorized interface that you are using is streamlining inserting into the
dictionary. 
> If you pad with zeros, you are inserting 85770 * 10 individual key/value
pairs, NOT 85770 individual key/value pairs. 7/10ths of your inserts are
going to overwrite the 0 key value.

I understand now :D I wrote my last comment a bit too fast before running the actual computation on the result of the query.

> To do what you want to do, you need to bitpack the XYZ coordinates to an
S16 using this datatype (0[int32], x[int32], y[int32], z[int32]) using the
nparray.view() function. Then you will need to store all of your values as
bitpacked bool arrays.

Thanks again, this is exactly what I need.

**Comment by atom-moyer** [comment]
Great! I'll add an example to the README of doing this. I do this often, so I have a lot of examples.

**Comment by atom-moyer** [comment]
You should be able to see an example now. If you have anymore questions, just reach out. Hopefully it works for you! Sounds like you are using it very similar to me.

**Comment by atom-moyer** [comment]
One last thing, I forgot about because I don't use it much.  There are "in-place" operators that I think you will need if you need to "add/modify" to your values. Basically, you can use the `Dict.ior()` method to flip bits.

    def iadd(self, key, value):
        if self.__safe_mode: assert key.dtype == self.__key_type and value.dtype == self.__value_type, (key.dtype, value.dtype)
        self.__dict.iadd(key, value)

    def isub(self, key, value):
        if self.__safe_mode: assert key.dtype == self.__key_type and value.dtype == self.__value_type, (key.dtype, value.dtype)
        self.__dict.isub(key, value)

    def ior(self, key, value):
        if self.__safe_mode: assert key.dtype == self.__key_type and value.dtype == self.__value_type, (key.dtype, value.dtype)
        self.__dict.ior(key, value)

    def iand(self, key, value):
        if self.__safe_mode: assert key.dtype == self.__key_type and value.dtype == self.__value_type, (key.dtype, value.dtype)
        self.__dict.iand(key, value)

**Comment by strohhut22** [comment]
> You should be able to see an example now. If you have anymore questions, just reach out. Hopefully it works for you! Sounds like you are using it very similar to me.

Yes, it does work now. I really appreciate your help! 

My final result for 3d coords + 10bit values:
```
key_type = np.dtype('S16')
value_type = np.dtype('S16')

vxls_temp = np.column_stack((np.zeros(len(vxls)), vxls)).astype(np.dtype('int32'))
keys = vxls_temp.view('S16')

bits_temp = tss.bool().numpy().astype(np.dtype(bool))
values = np.column_stack((bits_temp, np.zeros((len(bits_temp), 6)).astype(bool))).view('S16')

gp_dict = gp.Dict(key_type, value_type)
gp_dict[keys] = values
```

> One last thing, I forgot about because I don't use it much. There are "in-place" operators that I think you will need if you need to "add/modify" to your values. Basically, you can use the Dict.ior() method to flip bits.

Good tip, may need it in the future. But so far I really only need to store a big number of keys with their bit sequence and then run large queries that take too long without a parallel implementation. Your implementation here is actually the only "real" parallel hashmap I could find for Python. Just thanks again^^
